File: converters/typescript/src/registry.ts
import type { Converter } from './traits';
import type { Protocol } from './types';

/** 转换器注册表 */
export class ConverterRegistry {
  private converters = new Map<string, Converter>();

  /** 注册转换器 */
  register(converter: Converter): void {
    this.converters.set(converter.name(), converter);
  }

  /** 注销转换器 */
  unregister(name: string): boolean {
    return this.converters.delete(name);
  }

  /** 获取转换器 */
  get(name: string): Converter | undefined {
    return this.converters.get(name);
  }

  /** 按协议对查找转换器 */
  find(source: Protocol, target: Protocol): Converter | undefined {
    for (const converter of this.converters.values()) {
      if (converter.canConvert(source, target)) {
        return converter;
      }
    }
    return undefined;
  }

  /** 列出所有已注册的转换器名称 */
  list(): string[] {
    return Array.from(this.converters.keys());
  }

  /** 按源协议筛选 */
  bySourceProtocol(protocol: Protocol): Converter[] {
    return Array.from(this.converters.values()).filter(
      (c) => c.sourceProtocol() === protocol
    );
  }

  /** 按目标协议筛选 */
  byTargetProtocol(protocol: Protocol): Converter[] {
    return Array.from(this.converters.values()).filter(
      (c) => c.targetProtocol() === protocol
    );
  }

  /** 注册所有内置转换器 */
  registerDefaults(): void {
    // OpenAI Responses →
    const { OpenAiResponsesToAnthropicMessagesConverter } = require('./openai-responses-to-anthropic-messages');
    const { OpenAiResponsesToOpenAiCompletionsConverter } = require('./openai-responses-to-openai-completions');
    const { OpenAiResponsesToGoogleGeminiConverter } = require('./openai-responses-to-google-gemini');

    // OpenAI Completions →
    const { OpenAiCompletionsToAnthropicMessagesConverter } = require('./openai-completions-to-anthropic-messages');
    const { OpenAiCompletionsToOpenAiResponsesConverter } = require('./openai-completions-to-openai-responses');
    const { OpenAiCompletionsToGoogleGeminiConverter } = require('./openai-completions-to-google-gemini');

    // Anthropic Messages →
    const { AnthropicMessagesToOpenAiResponsesConverter } = require('./anthropic-messages-to-openai-responses');
    const { AnthropicMessagesToOpenAiCompletionsConverter } = require('./anthropic-messages-to-openai-completions');
    const { AnthropicMessagesToGoogleGeminiConverter } = require('./anthropic-messages-to-google-gemini');

    // Google Gemini →
    const { GoogleGeminiToAnthropicMessagesConverter } = require('./google-gemini-to-anthropic-messages');
    const { GoogleGeminiToOpenAiResponsesConverter } = require('./google-gemini-to-openai-responses');
    const { GoogleGeminiToOpenAiCompletionsConverter } = require('./google-gemini-to-openai-completions');

    this.register(new OpenAiResponsesToAnthropicMessagesConverter());
    this.register(new OpenAiResponsesToOpenAiCompletionsConverter());
    this.register(new OpenAiResponsesToGoogleGeminiConverter());
    this.register(new OpenAiCompletionsToAnthropicMessagesConverter());
    this.register(new OpenAiCompletionsToOpenAiResponsesConverter());
    this.register(new OpenAiCompletionsToGoogleGeminiConverter());
    this.register(new AnthropicMessagesToOpenAiResponsesConverter());
    this.register(new AnthropicMessagesToOpenAiCompletionsConverter());
    this.register(new AnthropicMessagesToGoogleGeminiConverter());
    this.register(new GoogleGeminiToAnthropicMessagesConverter());
    this.register(new GoogleGeminiToOpenAiResponsesConverter());
    this.register(new GoogleGeminiToOpenAiCompletionsConverter());
  }
}

File: converters/rust/src/registry.rs
use std::collections::HashMap;
use std::sync::Arc;

use crate::error::ConverterError;
use crate::traits::Converter;
use crate::types::Protocol;

/// 转换器注册表
pub struct ConverterRegistry {
    converters: HashMap<String, Arc<dyn Converter>>,
}

impl ConverterRegistry {
    pub fn new() -> Self {
        Self {
            converters: HashMap::new(),
        }
    }

    /// 注册转换器
    pub fn register(&mut self, converter: Arc<dyn Converter>) {
        let name = converter.name().to_string();
        self.converters.insert(name, converter);
    }

    /// 注销转换器
    pub fn unregister(&mut self, name: &str) -> Option<Arc<dyn Converter>> {
        self.converters.remove(name)
    }

    /// 获取转换器
    pub fn get(&self, name: &str) -> Option<&Arc<dyn Converter>> {
        self.converters.get(name)
    }

    /// 按协议对查找转换器
    pub fn find(&self, source: &Protocol, target: &Protocol) -> Option<&Arc<dyn Converter>> {
        self.converters
            .values()
            .find(|c| c.can_convert(source, target))
    }

    /// 列出所有已注册的转换器名称
    pub fn list(&self) -> Vec<&str> {
        self.converters.keys().map(|k| k.as_str()).collect()
    }

    /// 按源协议筛选
    pub fn by_source_protocol(&self, protocol: &Protocol) -> Vec<&Arc<dyn Converter>> {
        self.converters
            .values()
            .filter(|c| c.source_protocol() == *protocol)
            .collect()
    }

    /// 按目标协议筛选
    pub fn by_target_protocol(&self, protocol: &Protocol) -> Vec<&Arc<dyn Converter>> {
        self.converters
            .values()
            .filter(|c| c.target_protocol() == *protocol)
            .collect()
    }

    /// 执行转换请求
    pub async fn convert_request(
        &self,
        converter_name: &str,
        request: crate::types::ConversionRequest,
    ) -> Result<crate::types::ConversionRequest, ConverterError> {
        let converter = self
            .get(converter_name)
            .ok_or_else(|| ConverterError::InternalError(format!("Converter not found: {}", converter_name)))?;
        converter.convert_request(request).await
    }

    /// 执行转换响应
    pub async fn convert_response(
        &self,
        converter_name: &str,
        response: crate::types::ConversionResponse,
    ) -> Result<crate::types::ConversionResponse, ConverterError> {
        let converter = self
            .get(converter_name)
            .ok_or_else(|| ConverterError::InternalError(format!("Converter not found: {}", converter_name)))?;
        converter.convert_response(response).await
    }

    /// 注册所有内置转换器
    pub fn register_defaults(&mut self) {
        use crate::anthropic_messages_to_google_gemini::AnthropicMessagesToGoogleGeminiConverter;
        use crate::anthropic_messages_to_openai_completions::AnthropicMessagesToOpenAiCompletionsConverter;
        use crate::anthropic_messages_to_openai_responses::AnthropicMessagesToOpenAiResponsesConverter;
        use crate::google_gemini_to_anthropic_messages::GoogleGeminiToAnthropicMessagesConverter;
        use crate::google_gemini_to_openai_completions::GoogleGeminiToOpenAiCompletionsConverter;
        use crate::google_gemini_to_openai_responses::GoogleGeminiToOpenAiResponsesConverter;
        use crate::openai_completions_to_anthropic_messages::OpenAiCompletionsToAnthropicMessagesConverter;
        use crate::openai_completions_to_google_gemini::OpenAiCompletionsToGoogleGeminiConverter;
        use crate::openai_completions_to_openai_responses::OpenAiCompletionsToOpenAiResponsesConverter;
        use crate::openai_responses_to_anthropic_messages::OpenAiResponsesToAnthropicMessagesConverter;
        use crate::openai_responses_to_google_gemini::OpenAiResponsesToGoogleGeminiConverter;
        use crate::openai_responses_to_openai_completions::OpenAiResponsesToOpenAiCompletionsConverter;

        // OpenAI Responses →
        self.register(Arc::new(OpenAiResponsesToAnthropicMessagesConverter::default()));
        self.register(Arc::new(OpenAiResponsesToOpenAiCompletionsConverter::default()));
        self.register(Arc::new(OpenAiResponsesToGoogleGeminiConverter::default()));

        // OpenAI Completions →
        self.register(Arc::new(OpenAiCompletionsToAnthropicMessagesConverter::default()));
        self.register(Arc::new(OpenAiCompletionsToOpenAiResponsesConverter::default()));
        self.register(Arc::new(OpenAiCompletionsToGoogleGeminiConverter::default()));

        // Anthropic Messages →
        self.register(Arc::new(AnthropicMessagesToOpenAiResponsesConverter::default()));
        self.register(Arc::new(AnthropicMessagesToOpenAiCompletionsConverter::default()));
        self.register(Arc::new(AnthropicMessagesToGoogleGeminiConverter::default()));

        // Google Gemini →
        self.register(Arc::new(GoogleGeminiToAnthropicMessagesConverter::default()));
        self.register(Arc::new(GoogleGeminiToOpenAiResponsesConverter::default()));
        self.register(Arc::new(GoogleGeminiToOpenAiCompletionsConverter::default()));
    }
}

impl Default for ConverterRegistry {
    fn default() -> Self {
        let mut registry = Self::new();
        registry.register_defaults();
        registry
    }
}
